janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.janusgraph.graphdb.util;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.JanusGraphException;
+import org.janusgraph.diskstorage.BackendTransaction;
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.query.profile.QueryProfiler;
+
+import com.google.common.cache.Cache;
+
+/**
+ * @author davidclement90@laposte.net
+ */
+public class SubqueryIterator implements Iterator<JanusGraphElement>, AutoCloseable {
+
+    private final JointIndexQuery.Subquery subQuery;
+
+    private final Cache<JointIndexQuery.Subquery, List<Object>> indexCache;
+
+    private Iterator<? extends JanusGraphElement> itElement;
+
+    private List<Object> currentIds;
+
+    private QueryProfiler profiler;
+
+    private boolean isTimerRunning;
+
+    public SubqueryIterator(JointIndexQuery.Subquery subQuery, IndexSerializer indexSerializer, BackendTransaction tx,
+            Cache<JointIndexQuery.Subquery, List<Object>> indexCache, int limit,
+            Function<Object, ? extends JanusGraphElement> function, List<Object> otherResults) {
+        this.subQuery = subQuery;
+        this.indexCache = indexCache;
+        List<Object> cacheResponse;
```

Comment:
`final` here and throughout PR

Reply:
This one is not final because _cacheResponse_ is set on line 59 but for all the others and even more I update my eclipse save actions.
We need to work on https://github.com/JanusGraph/janusgraph/issues/242...